src/test/neorados/test_neorados.cc

```diff
@@ -29,11 +54,484 @@ TEST_F(TestNeoRADOS, MakeWithLibRADOS) {
+
+      co_await rados.delete_pool(pool_id, 
+          boost::asio::redirect_error(boost::asio::use_awaitable, ec));
+      if (ec) throw boost::system::system_error(ec);
+      
+    }, [](std::exception_ptr e) {
+      if (e) std::rethrow_exception(e);
+    }
+  );
+
+  ASSERT_NO_THROW(ctx.run());
+  ASSERT_GT(pool, 0);
+}
+
+TEST_F(TestNeoRADOSCoro, CreateAndListObjects) {
+  boost::asio::io_context ctx;
+  boost::system::error_code ec;
+  
+  std::set<std::string> objects;
+  std::set<std::string> fetched_objects;
+  fill_objects(1, 100, objects);
+
+  boost::asio::co_spawn(ctx, 
+    [&]() -> boost::asio::awaitable<void> {
+      auto rados = co_await RADOS::Builder{}.build(ctx, boost::asio::use_awaitable);
+      const std::string pool_id{"piscine"};
+      auto pool_deleter = make_scope_guard([&]() {
+          rados.delete_pool(pool_id, boost::asio::detached);
+        });
+      co_await rados.create_pool(pool_id, std::nullopt, 
+          boost::asio::redirect_error(boost::asio::use_awaitable, ec));
+      if (ec) throw boost::system::system_error(ec);
+
+      const auto pool = co_await rados.lookup_pool(pool_id, 
+          boost::asio::redirect_error(boost::asio::use_awaitable, ec));
+      if (ec) throw boost::system::system_error(ec);
+
+      IOContext io_ctx(pool);
+      for (const auto& o : objects) {
+        WriteOp op;
+        ceph::bufferlist bl;
+        bl.append("there is nothing to see here");
+        op.write_full(std::move(bl));
+        co_await rados.execute(o, io_ctx, std::move(op),
+            boost::asio::redirect_error(boost::asio::use_awaitable, ec));
+        if (ec) throw boost::system::system_error(ec);
+      }
+
+      auto b = Cursor::begin();
+      const auto e = Cursor::end();
+      std::vector<Entry> v;
+      Cursor next = Cursor::begin();
+      do {
+        std::tie(v, next) = co_await rados.enumerate_objects(io_ctx, b, e, 1000, {},
+          boost::asio::redirect_error(boost::asio::use_awaitable, ec));
+        if (ec) throw boost::system::system_error(ec);
+        std::transform(v.cbegin(), v.cend(), 
+          std::inserter(fetched_objects, fetched_objects.begin()), 
+          [](const Entry& e){return e.oid;}
+        );
+      } while (next != Cursor::end());
+
+    }, [](std::exception_ptr e) {
+      if (e) std::rethrow_exception(e);
+    }
+  );
+
+  ASSERT_NO_THROW(ctx.run());
+  ASSERT_EQ(fetched_objects, objects);
+}
+
+TEST_F(TestNeoRADOSCoro, ListPools) {
+  boost::asio::io_context ctx;
+  boost::system::error_code ec;
+
+  const std::set<std::string> pools{"piscine", "piscina", "zwembad"};
+  std::set<std::string> fetched_pools;
+
+  boost::asio::co_spawn(ctx, 
+    [&]() -> boost::asio::awaitable<void> {
+      auto rados = co_await RADOS::Builder{}.build(ctx, boost::asio::use_awaitable);
+      auto pool_deleter = make_scope_guard([&]() {
+          for (const auto& pool: pools) {
+            rados.delete_pool(pool, boost::asio::detached);
```

Comment:
this seems to delete only one of the pools

Reply:
Yes. I think fundamentally the problem is that once the destructor exits things fundamentally get deleted and without the destructor being a coroutine there's no good way to wait, especially in a single threaded model.

I've got https://github.com/ceph/ceph/pull/48234/commits/3c3732625a6afdaefdd213b9dd2e27621ce53843#diff-e7a05e42b0b2a31b55db86bb74d1c3cfe3c664730c8b939383a90ca945f0ada9R65 in progress that I think might do a better job for coroutine tests.

Reply:
my concern is not about the unit tests here (for that I can just delete all pools in blocking calls after the test).
the main goal is to understand how to write exception-safe code with coroutines.

Reply:
by the time we get to an object's destructor, it's too late for any async cleanup. RAII works fine for stuff like memory allocations, but any async cleanup needs to be part of the async control flow

for graceful shutdowns, maybe take a look at the asio frontend for an example? in https://github.com/ceph/ceph/blob/main/src/rgw/rgw_asio_frontend.cc#L1063-L1091 we have a `stop()` function that closes sockets, and a `join()` function that waits for `io_context.run()` to complete on all of its threads - that drains the `io_context` work queue. if `stop()` happens in the middle of a call to `process_request()`, that call will keep running until it hits a socket error, then follow its normal error path until the coroutine completes

in contrast, you might instead call `io_context.stop()` to cause all of the `io_context.run()` threads to return immediately without draining. here we still need to rely on RAII to guarantee that all of the memory gets released, but we won't have a chance to do any async cleanup. this was the strategy you took in https://github.com/ceph/ceph/blob/main/src/rgw/rgw_notify.cc#L467-L470 and i think it works fine there